discovery-provider/src/tasks/index_spl_token.py

```diff
@@ -0,0 +1,451 @@
+        or "message" not in tx_result["transaction"]
+        or "accountKeys" not in tx_result["transaction"]["message"]
+    ):
+        logger.info("iomnvalid format, return early")
+        return (root_accounts_to_refresh, token_accounts_to_refresh)
+    owner_balance_dict = {}
+    account_keys = tx_result["transaction"]["message"]["accountKeys"]
+    for pre_balance in tx_result["meta"]["preTokenBalances"]:
+        owner_balance_dict[pre_balance["accountIndex"]] = pre_balance["uiTokenAmount"][
+            "amount"
+        ]
+    for post_balance in tx_result["meta"]["postTokenBalances"]:
+        if (
+            post_balance["accountIndex"] not in owner_balance_dict
+            and post_balance["uiTokenAmount"]["amount"] != "0"
+        ):
+            root_accounts_to_refresh.add(post_balance["owner"])
+            token_accounts_to_refresh.add(account_keys[post_balance["accountIndex"]])
+        elif (
+            post_balance["accountIndex"] in owner_balance_dict
+            and post_balance["uiTokenAmount"]["amount"]
+            != owner_balance_dict[post_balance["accountIndex"]]
+        ):
+            root_accounts_to_refresh.add(post_balance["owner"])
+            token_accounts_to_refresh.add(account_keys[post_balance["accountIndex"]])
+    return (root_accounts_to_refresh, token_accounts_to_refresh)
+
+
+# Cache the latest value committed to DB in redis
+# Used for quick retrieval in health check
+def cache_latest_spl_audio_db_tx(redis: Redis, latest_tx: CachedProgramTxInfo):
+    cache_latest_sol_db_tx(redis, latest_sol_spl_token_db_key, latest_tx)
+
+
+def parse_spl_token_transaction(
+    solana_client_manager: SolanaClientManager,
+    tx_sig: ConfirmedSignatureForAddressResult,
+) -> Tuple[ConfirmedTransaction, List[str], List[str]]:
+    try:
+        tx_info = solana_client_manager.get_sol_tx_info(tx_sig["signature"])
+        result = tx_info["result"]
+        error = tx_info["result"]["meta"]["err"]
+
+        if error:
+            return (tx_info, [], [])
+        root_accounts, token_accounts = get_token_balance_change_owners(result)
+        return (tx_info, list(root_accounts), list(token_accounts))
+
+    except Exception as e:
+        signature = tx_sig["signature"]
+        logger.error(
+            f"index_spl_token.py | Error processing {signature}, {e}", exc_info=True
+        )
+        raise e
+
+
+# Query the highest traversed solana slot
+def get_latest_slot(db):
+    latest_slot = None
+    with db.scoped_session() as session:
+        highest_slot_query = session.query(
+            SPLTokenTransaction.last_scanned_slot
+        ).first()
+        # Can be None prior to first write operations
+        if highest_slot_query is not None:
+            latest_slot = highest_slot_query[0]
+
+    # Return None if not yet cached
+
+    logger.info(f"index_spl_token.py | returning {latest_slot} for highest slot")
+    return latest_slot
+
+
+def parse_sol_tx_batch(
```

Comment:
Let's just sanity check that we have proper indices for the queries in the fn (esp the associated wallets)

Reply:
there is an index on (wallet, is_current)
seems really quick - there are only 70k rows in this table and 17k which are is_current

![Screen Shot 2022-03-04 at 7 45 46 AM](https://user-images.githubusercontent.com/7064122/156765973-9a9007c7-a5a8-4bfc-bd71-5751e9e95147.png)